src/pybind/mgr/dashboard/run-backend-api-tests.sh

```diff
@@ -93,14 +92,19 @@ EOF
     cd $CURR_DIR
 
     COVERAGE_VERSION=$(cat requirements.txt | grep coverage)
+    COVERAGE_INSTALL_PATH=$(mktemp -d)
```

Comment:
i don't quite understand this change. is this only for testing?

Reply:
> i don't quite understand this change. is this only for testing?

@tchaikov This is the rationale: to measure coverage we need ```coverage``` package:
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/module.py#L63

But we want:
1) coverage NOT to be installed by system pip, but venv's pip.
2) Add to PYTHONPATH the path to find coverage only, not the whole virtualenv site-packages (that could have side-effects: there are lot of packages that ceph-dashboard module shouldn't be aware of).

So I decided to install coverage through virtualenv but in isolated directory to avoid side effects.

Reply:
> coverage NOT to be installed by system pip, but venv's pip.

i think i helped to root cause this issue yesterday. and as an option i suggested to move `deactivate` down after we've done with the venv. that's what we were discussing yesterday. 

> Add to PYTHONPATH the path to find coverage only, not the whole virtualenv site-packages (that could have side-effects: there are lot of packages that ceph-dashboard module shouldn't be aware of).

i don't understand this part. the purpose of `setup_teuthology()` is to setup a venv. the whole point of this very venv is for preparing the packages for following line
```
eval python ../qa/tasks/vstart_runner.py --skip-binaries-checking $OPTIONS $TEST_CASES
```

does your change make any difference?

Reply:
> i don't understand this part. the purpose of `setup_teuthology()` is to setup a venv. the whole point of this very venv is for preparing the packages for following line
> 
> ```
> eval python ../qa/tasks/vstart_runner.py --skip-binaries-checking $OPTIONS $TEST_CASES
> ```
> 
> does your change make any difference?

The package ```coverage``` needs to be discoverable by dashboard module, so we need it in the PYTHONPATH or the module crashes when env. var COVERAGE_ENABLED=true.

So we install it in the same venv as teuthology in order to avoid installing it in the system (and not in another venv dedicated only to coverage). Because of this we moved down ```deactivate```.
I put it in a separate directory to avoid the other packages in venv to be in the PYTHONPATH (so they are not discoverable by dashboard module to avoid side effects).

```coverage``` package is only needed in this script to measure tests coverage:
```
Name                                                                  Stmts   Miss  Cover
-----------------------------------------------------------------------------------------
/ceph/src/pybind/mgr/dashboard/__init__.py                               33     29    12%
/ceph/src/pybind/mgr/dashboard/awsauth.py                                80     37    54%
/ceph/src/pybind/mgr/dashboard/controllers/__init__.py                  545    472    13%
/ceph/src/pybind/mgr/dashboard/controllers/auth.py                       50     34    32%
/ceph/src/pybind/mgr/dashboard/controllers/cephfs.py                    140     55    61%
/ceph/src/pybind/mgr/dashboard/controllers/cluster_configuration.py      52     16    69%
/ceph/src/pybind/mgr/dashboard/controllers/erasure_code_profile.py       38     17    55%
/ceph/src/pybind/mgr/dashboard/controllers/health.py                    124     36    71%
/ceph/src/pybind/mgr/dashboard/controllers/host.py                        8      7    13%
/ceph/src/pybind/mgr/dashboard/controllers/logs.py                       32     22    31%
/ceph/src/pybind/mgr/dashboard/controllers/mgr_modules.py                77     31    60%
/ceph/src/pybind/mgr/dashboard/controllers/monitor.py                    24     11    54%
/ceph/src/pybind/mgr/dashboard/controllers/nfsganesha.py                123     79    36%
/ceph/src/pybind/mgr/dashboard/controllers/osd.py                       107     41    62%
/ceph/src/pybind/mgr/dashboard/controllers/perf_counters.py              48     30    38%
/ceph/src/pybind/mgr/dashboard/controllers/pool.py                      115     40    65%
/ceph/src/pybind/mgr/dashboard/controllers/rbd.py                       318    103    68%
/ceph/src/pybind/mgr/dashboard/controllers/rbd_mirroring.py             256    180    30%
/ceph/src/pybind/mgr/dashboard/controllers/rgw.py                       191     84    56%
/ceph/src/pybind/mgr/dashboard/controllers/role.py                       77     21    73%
/ceph/src/pybind/mgr/dashboard/controllers/settings.py                   36     20    44%
/ceph/src/pybind/mgr/dashboard/controllers/summary.py                    60     39    35%
/ceph/src/pybind/mgr/dashboard/controllers/task.py                        8      6    25%
/ceph/src/pybind/mgr/dashboard/controllers/user.py                       69     21    70%
/ceph/src/pybind/mgr/dashboard/exceptions.py                             53     32    40%
/ceph/src/pybind/mgr/dashboard/module.py                                247    234     5%
/ceph/src/pybind/mgr/dashboard/plugins/__init__.py                       33     31     6%
/ceph/src/pybind/mgr/dashboard/plugins/feature_toggles.py                70     63    10%
/ceph/src/pybind/mgr/dashboard/plugins/pluggy.py                         38     36     5%
/ceph/src/pybind/mgr/dashboard/rest_client.py                           259    127    51%
/ceph/src/pybind/mgr/dashboard/security.py                               33     29    12%
/ceph/src/pybind/mgr/dashboard/services/access_control.py               395    294    26%
/ceph/src/pybind/mgr/dashboard/services/auth.py                         123     43    65%
/ceph/src/pybind/mgr/dashboard/services/ceph_service.py                 147     43    71%
/ceph/src/pybind/mgr/dashboard/services/cephfs.py                        58     33    43%
/ceph/src/pybind/mgr/dashboard/services/cephx.py                         19      8    58%
/ceph/src/pybind/mgr/dashboard/services/exception.py                     82     46    44%
/ceph/src/pybind/mgr/dashboard/services/ganesha.py                      648    230    65%
/ceph/src/pybind/mgr/dashboard/services/orchestrator.py                  37     31    16%
/ceph/src/pybind/mgr/dashboard/services/rbd.py                           69     25    64%
/ceph/src/pybind/mgr/dashboard/services/rgw_client.py                   156     70    55%
/ceph/src/pybind/mgr/dashboard/services/tcmu_service.py                  57     54     5%
/ceph/src/pybind/mgr/dashboard/settings.py                              100     92     8%
/ceph/src/pybind/mgr/dashboard/tools.py                                 530    253    52%
-----------------------------------------------------------------------------------------
TOTAL                                                                  5765   3205    44%
```